src/lib.rs

```diff
@@ -7,76 +7,81 @@
 //! ```
-use std::{
-    debug_assert,
-    io::{self, Read, Write},
-    time::Duration,
-};
+use std::debug_assert;
+use std::io::{self, Read, Write};
+use std::time::Duration;
 
-const ACCEPTABLE_SPEED_DIFF: f64 = 4.0 / 100.0;
+#[cfg(test)]
+mod tests;
 
 #[derive(Clone, Debug)]
 pub struct LimiterOptions {
     pub window_length: u64,
     pub window_time: Duration,
     pub bucket_size: u64,
+    pub min_operation_size: u64,
+
+    // Store constants based on options to avoid re-computation at runtime
+    pub tsleep: Duration,
+    pub wtime_ns: u64,
+    pub stream_cap_limit: u64,
+    pub sleep_threshold: u64,
 }
 
 impl LimiterOptions {
     pub fn new(
         mut window_length: u64,
         mut window_time: Duration,
-        mut bucket_size: u64,
+        bucket_size: u64,
     ) -> LimiterOptions {
-        let rate = window_length.min(bucket_size) as f64;
-        let tw: f64 = window_time.as_nanos() as f64;
-        let init_speed = (rate / tw) * 1_000_000.0;
-
-        let mut new_speed = init_speed;
-        let mut new_wlen = window_length;
-        let mut new_wtime = window_time;
-        let mut new_bsize = bucket_size;
-
-        // While the difference between the intented speed (init_speed) and the reduced one (new_speed) is under the threshold
-        //    Each iteration, divide all the options by 2, and recompute the speed (in order to check if it's not altered)
-        //    Because we want the values BEFORE the speed is above the threshold, assign the new values on start of the new iter only
-        while ((new_speed / init_speed) - 1.0).abs() < ACCEPTABLE_SPEED_DIFF {
-            // Values from past iter, we know they're under the threshold
-            window_length = new_wlen;
-            window_time = new_wtime;
-            bucket_size = new_bsize;
-
-            // If values aren't dividable by 2
-            if (new_wlen == 1) || (new_bsize == 1) || (new_wtime.as_nanos() == 1) {
-                break;
-            }
-
-            // Reduce the options
-            new_wlen /= 2;
-            new_wtime /= 2;
-            new_bsize /= 2;
+        let mut wlen = window_length;
+        let mut wtime = window_time;
+        while wlen > u32::MAX as u64 {
+            wlen /= 2;
+            wtime /= 2;
+        }
+        let tsleep = wtime / TryInto::<u32>::try_into(wlen).unwrap();
 
-            // Recompute the new speed
-            let rate = new_wlen.min(new_bsize) as f64;
-            let tw: f64 = new_wtime.as_nanos() as f64;
-            new_speed = (rate / tw) * 1_000_000.0;
+        while window_time.as_nanos() > u64::MAX as u128 {
+            window_time /= 2;
+            window_length /= 2;
         }
 
+        let stream_cap_limit = std::cmp::min(window_length, bucket_size);
         LimiterOptions {
+            stream_cap_limit,
+            wtime_ns: window_time.as_nanos().try_into().unwrap(),
+            sleep_threshold: stream_cap_limit.min(bucket_size),
```

Comment:
`stream_cap_limit` is already a min with `bucket_size` L.57 don't need to do it again.